### PR TITLE
first try solution

### PR DESCRIPTION
**Pull request by AndreyDMA**

### CODE REVIEW

File: src/main/java/core/basesyntax/ArrayList.java [hunk 1]
```diff
@@ -1,52 +1,147 @@
 package core.basesyntax;
 
-/**
- * <p>Реалізувати свій ArrayList який імплементує інтерфейс List</p>
- */
 public class ArrayList<T> implements List<T> {
+    private int size;
+    private Object[] array = new Object[10];
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: 10 is the magic number?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 2]
```diff
@@ -1,52 +1,147 @@
...
+    private Object[] tempArray = new Object[10];
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: do we need tempArray here? try don't use it

File: src/main/java/core/basesyntax/ArrayList.java [hunk 3]
```diff
@@ -1,52 +1,147 @@
...
 
     @Override
     public void add(T value) {
-
+        if (size >= array.length) {
+            tempArray = new Object[array.length];
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: try to use method System.arraucopy()
and better move this logic in separate private method witch will increase size of the array

File: src/main/java/core/basesyntax/ArrayList.java [hunk 4]
```diff
@@ -1,52 +1,147 @@
...
+            for (int i = 0; i < array.length; i++) {
+                tempArray[i] = array[i];
+            }
+            array = new Object[size * 2];
+            for (int i = 0; i < tempArray.length; i++) {
+                array[i] = tempArray[i];
+            }
+        }
+        array[size] = value;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        array[size++] = value;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 5]
```diff
@@ -1,52 +1,147 @@
...
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: try to simplify this methof

File: src/main/java/core/basesyntax/ArrayList.java [hunk 6]
```diff
@@ -1,52 +1,147 @@
...
+        for (int i = 0; i < list.size(); i++) {
+            array[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return (T) array[index];
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: add checking for index

File: src/main/java/core/basesyntax/ArrayList.java [hunk 7]
```diff
@@ -1,52 +1,147 @@
...
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: simplify this method

File: src/main/java/core/basesyntax/ArrayList.java (hunk 1 above)
Comment (AndreyDMA) [reply src/main/java/core/basesyntax/ArrayList.java]: 10 is default capacity of array in ArrayList...

File: src/main/java/core/basesyntax/ArrayList.java [hunk 8]
```diff
@@ -1,52 +1,120 @@
...
+            System.arraycopy(objects, 0, resObjects, 0, size);
+            objects = resObjects;
+        }
+    }
 
     @Override
     public void add(T value) {
-
+        resize();
+        objects[size] = value;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]:
```suggestion
        objects[size++] = value;
```

File: src/main/java/core/basesyntax/ArrayList.java [hunk 9]
```diff
@@ -1,52 +1,120 @@
...
+        resize();
+        System.arraycopy(objects, index, objects, index + 1, size - index);
+        objects[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (size + list.size() >= objects.length) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: maybe while() ?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 10]
```diff
@@ -1,52 +1,120 @@
...
+        for (int i = 0; i < list.size(); i++) {
+            objects[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return (T) objects[index];
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: add checking for index

File: src/main/java/core/basesyntax/ArrayList.java [hunk 11]
```diff
@@ -1,52 +1,120 @@
...
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index > size) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: if index <0?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 12]
```diff
@@ -1,52 +1,120 @@
...
+            throw new ArrayIndexOutOfBoundsException("java.lang.ArrayIndexOutOfBoundsException");
+        }
+        objects[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index > size || size < 0) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: size < 0 or index< 0

File: src/main/java/core/basesyntax/ArrayList.java [hunk 13]
```diff
@@ -1,52 +1,120 @@
...
+            throw new ArrayIndexOutOfBoundsException("java.lang.ArrayIndexOutOfBoundsException");
+        }
+        Object[] tempArray = new Object[objects.length];
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: can we make it without temp array?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 14]
```diff
@@ -1,52 +1,120 @@
...
+            }
+        }
+        size--;
+        return (T) tempArray[index];
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        if (t == null) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: try to use method remove (int index) here

File: src/main/java/core/basesyntax/ArrayList.java [hunk 15]
```diff
@@ -1,52 +1,120 @@
...
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size < 0;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: does this method work correct?

**Comment by AndreyDMA** [comment]
Check pls, I think I fix everything...

File: src/main/java/core/basesyntax/ArrayList.java [hunk 16]
```diff
@@ -33,79 +32,64 @@ public void add(T value, int index) {
...
-            if (i < index) {
-                objects[i] = tempArray[i];
-            } else {
-                objects[i] = tempArray[i + 1];
-            }
-        }
+        System.arraycopy(objects, index + 1, objects, index, objects.length - index - 1);
         size--;
-        return (T) tempArray[index];
+        if (size < 0) {
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: maybe better move this `if` statement before System.arraycopy?

File: src/main/java/core/basesyntax/ArrayList.java [hunk 17]
```diff
@@ -1,52 +1,104 @@
...
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        if (t == null) {
+            throw new ArrayIndexOutOfBoundsException("java.lang.ArrayIndexOutOfBoundsException");
+        }
+        //Object[] tempArray = new Object[objects.length];
+        int count = 0;
```
Comment (Sofasmile) [inline src/main/java/core/basesyntax/ArrayList.java]: for (int i = 0; i < objects.length; i++) {
            if (t.equals(objects[i])) {
                return remove(i);
            }
        }
            throw new ArrayIndexOutOfBoundsException("java.lang.ArrayIndexOutOfBoundsException");

**Comment by AndreyDMA** [comment]
Sorry for my code style, I start my self learning only from May... So I understood your displeasure...

**Pull request closed without merging**